File: components/display.py
import subprocess
import re
import os
from PyQt6.QtWidgets import QLabel, QWidget, QComboBox, QSlider, QPushButton, QMessageBox
from PyQt6.QtGui import QPixmap, QFont
from PyQt6.QtCore import Qt, QTimer


class DisplayManager:
    def __init__(self, parent):
        self.parent = parent
        self.refresh_timer = QTimer()
        self.refresh_timer.setInterval(1000)
        self.refresh_timer.timeout.connect(self.auto_refresh)
        self.init_display()

    def detect_theme_image(self):
        base_path = os.path.expanduser("~/Elysia/assets/")
        try:
            output = subprocess.check_output(
                ["gsettings", "get", "org.gnome.desktop.interface", "color-scheme"]
            ).decode().strip()
            return os.path.join(base_path, "assets/light.png") if "prefer-light" in output else os.path.join(base_path, "assets/dark.png")
        except Exception:
            return os.path.join(base_path, "assets/light.png")

    def init_display(self):
        self.desktop_label = QLabel("", self.parent)
        self.desktop_label.setFont(QFont("ElysiaOSNew", 16, QFont.Weight.Bold))
        self.desktop_label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
        self.desktop_label.move(760, 0)
        self.desktop_label.setVisible(False)

        self.desktop_image_label = QLabel(self.parent)
        self.desktop_image_label.setGeometry(650, 45, 300, 335)
        self.desktop_image_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self.desktop_image_label.setVisible(False)

        self.display_box = QWidget(self.parent)
        self.display_box.setGeometry(500, 100, 600, 500)
        self.display_box.setStyleSheet("""
            background-color: rgba(255, 255, 255, 0.05);
            border: 2px solid rgba(255, 255, 255, 0.15);
            border-radius: 30px;
        """)
        self.display_box.setVisible(False)

        self.resolution_label = QLabel("Resolution", self.display_box)
        self.resolution_label.setFont(QFont("ElysiaOSNew", 14))
        self.resolution_label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
        self.resolution_label.move(220, 205)

        self.display_dropdown = QComboBox(self.display_box)
        self.display_dropdown.setGeometry(150, 240, 300, 40)
        self.display_dropdown.setStyleSheet("""
            QComboBox {
                background: rgba(255, 255, 255, 0.2);
                color: white;
                border: 1px solid rgba(255, 255, 255, 0.25);
                border-radius: 21px;
                padding: 5px;
                font-size: 16px;
            }
            QComboBox:hover {
                background: rgba(255, 255, 255, 0.3);
            }
            QComboBox QAbstractItemView {
                background: rgba(0, 0, 0, 0.2);
                border: none;
                color: white;
                border-radius: 21px;
                padding: 5px;
                selection-background-color: rgba(255, 255, 255, 0.2);
            }
        """)

        self.brightness_label = QLabel("Brightness", self.display_box)
        self.brightness_label.setFont(QFont("ElysiaOSNew", 14))
        self.brightness_label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
        self.brightness_label.move(220, 300)

        self.brightness_slider = QSlider(Qt.Orientation.Horizontal, self.display_box)
        self.brightness_slider.setGeometry(100, 340, 400, 30)
        self.brightness_slider.setMinimum(1)
        self.brightness_slider.setMaximum(100)
        self.brightness_slider.setStyleSheet("""
            QSlider::horizontal {
                background: transparent;
                border: none;
                padding: 0px;
                margin: 0px;
            }
            QSlider::groove:horizontal {
                height: 8px;
                background: transparent;
                border: none;
                padding: 0px;
                margin: 0px;
            }
            QSlider::handle:horizontal {
                background: pink;
                border: 1px solid white;
                width: 18px;
                margin: -6px 0;
                border-radius: 9px;
            }
            QSlider::sub-page:horizontal {
                background: rgba(255, 255, 255, 0.25);
                border-radius: 5px;
            }
            QSlider::add-page:horizontal {
                background: rgba(255, 255, 255, 0.05);
                border-radius: 5px;
            }
        """)
        self.brightness_slider.valueChanged.connect(self.set_brightness)

        self.apply_btn = QPushButton("Apply Changes", self.display_box)
        self.apply_btn.setGeometry(200, 400, 200, 50)
        self.apply_btn.setStyleSheet("""
            QPushButton {
                background-color: #e5a7c6;
                color: white;
                font-weight: bold;
                border: none;
                padding: 6px 12px;
                letter-spacing: 1px;
                border-radius: 10px;
                font-size: 16px;
            }
            QPushButton:hover {
                background-color: #edcee3;
            }
        """)
        self.apply_btn.clicked.connect(self.apply_display_change)

    def show(self):
        self.update_display_dropdown()
        self.update_brightness_slider()
        self.update_theme_image()

        self.desktop_image_label.setVisible(True)
        self.desktop_label.setVisible(True)
        self.display_box.setVisible(True)

        self.parent.hide_tiles()
        self.parent.show_back_button()

        self.refresh_timer.start()

    def hide(self):
        self.refresh_timer.stop()
        self.display_box.setVisible(False)
        self.desktop_image_label.setVisible(False)
        self.desktop_label.setVisible(False)
        self.parent.hide_back_button()
        self.parent.show_tiles()

    def auto_refresh(self):
        if self.display_box.isVisible():
            self.update_brightness_slider()
            self.update_display_dropdown()
            self.update_theme_image()

    def update_display_dropdown(self):
        modes = self.get_monitor_modes()
        current_items = [self.display_dropdown.itemText(i) for i in range(self.display_dropdown.count())]
        if set(modes) != set(current_items):
            self.display_dropdown.clear()
            self.display_dropdown.addItems(modes)

        current_mode = None
        try:
            with open("/home/matsuko/.config/hypr/hyprland.conf", "r") as f:
                for line in f:
                    if line.startswith("monitor") and "@" in line:
                        match = re.search(r"\d{3,4}x\d{3,4}@\d{2,3}", line)
                        if match:
                            current_mode = match.group(0)
                        self.parent.monitor_name = line.split()[1]
                        break
        except Exception:
            self.parent.monitor_name = None

        if current_mode and current_mode in modes:
            self.display_dropdown.setCurrentText(current_mode)

    def update_theme_image(self):
        image_path = self.detect_theme_image()
        self.desktop_image_label.setPixmap(QPixmap(image_path).scaled(
            300, 500, Qt.AspectRatioMode.KeepAspectRatio, Qt.TransformationMode.SmoothTransformation))

    def get_monitor_modes(self):
        try:
            output = subprocess.check_output(["hyprctl", "monitors"], stderr=subprocess.STDOUT).decode()
        except Exception:
            return []
        return list(sorted(set(re.findall(r"\d{3,4}x\d{3,4}@\d{2,3}", output))))

    def set_brightness(self, value):
        try:
            subprocess.run(["brightnessctl", "set", f"{value}%"], check=True)
        except Exception as e:
            QMessageBox.warning(self.parent, "Error", f"Failed to set brightness:\n{e}")

    def update_brightness_slider(self):
        try:
            output = subprocess.check_output(["brightnessctl", "g"]).decode().strip()
            max_output = subprocess.check_output(["brightnessctl", "m"]).decode().strip()
            percent = int(int(output) / int(max_output) * 100)
            if self.brightness_slider.value() != percent:
                self.brightness_slider.setValue(percent)
        except Exception:
            self.brightness_slider.setValue(50)

    def get_current_monitor_line(self):
        path = "/home/matsuko/.config/hypr/hyprland.conf"
        try:
            with open(path, "r") as f:
                lines = f.readlines()
        except FileNotFoundError:
            return None, []

        for i, line in enumerate(lines):
            if line.startswith("monitor") and "@" in line:
                return i, lines
        return None, lines

    def apply_display_change(self):
        new_mode = self.display_dropdown.currentText()
        index, lines = self.get_current_monitor_line()
        if index is None:
            QMessageBox.warning(self.parent, "Error", "Could not find monitor config line.")
            return

        lines[index] = re.sub(r"\d{3,4}x\d{3,4}@\d{2,3}", new_mode, lines[index])
        try:
            with open("/home/matsuko/.config/hypr/hyprland.conf", "w") as f:
                f.writelines(lines)
        except Exception as e:
            QMessageBox.warning(self.parent, "Error", f"Failed to write file:\n{e}")
            return

        QMessageBox.information(self.parent, "Success", f"Display mode changed to {new_mode}.\nRestart Hyprland for changes to apply.")

File: components/Display.cpp
#include "Display.h"
#include <QDir>
#include <QFont>
#include <QPixmap>
#include <QProcess>
#include <QMessageBox>
#include <QFile>
#include <QTextStream>
#include <QRegularExpression>
#include <QRegularExpressionMatch>

DisplayManager::DisplayManager(QObject *parent) : QObject(parent) {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent);
    if (!parentWidget) return;
    
    refreshTimer = new QTimer(this);
    refreshTimer->setInterval(5000);
    connect(refreshTimer, &QTimer::timeout, this, &DisplayManager::autoRefresh);
    
    // Desktop label
    desktopLabel = new QLabel("", parentWidget);
    QFont desktopFont("ElysiaOSNew", 16, QFont::Bold);
    desktopLabel->setFont(desktopFont);
    desktopLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
    desktopLabel->move(760, 0);
    desktopLabel->setVisible(false);
    
    // Desktop image label
    desktopImageLabel = new QLabel(parentWidget);
    desktopImageLabel->setGeometry(650, 45, 300, 335);
    desktopImageLabel->setAlignment(Qt::AlignCenter);
    desktopImageLabel->setVisible(false);
    
    // Display box
    displayBox = new QWidget(parentWidget);
    displayBox->setGeometry(500, 100, 600, 500);
    displayBox->setStyleSheet(
        "background-color: rgba(255, 255, 255, 0.05);"
        "border: 2px solid rgba(255, 255, 255, 0.15);"
        "border-radius: 30px;"
    );
    displayBox->setVisible(false);
    
    // Resolution label
    resolutionLabel = new QLabel("Resolution", displayBox);
    QFont resFont("ElysiaOSNew", 14);
    resolutionLabel->setFont(resFont);
    resolutionLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
    resolutionLabel->move(220, 205);
    
    // Display dropdown
    displayDropdown = new QComboBox(displayBox);
    displayDropdown->setGeometry(150, 240, 300, 40);
    displayDropdown->setStyleSheet(
        "QComboBox {"
        "    background: rgba(255, 255, 255, 0.2);"
        "    color: white;"
        "    border: 1px solid rgba(255, 255, 255, 0.25);"
        "    border-radius: 21px;"
        "    padding: 5px;"
        "    font-size: 16px;"
        "}"
        "QComboBox:hover {"
        "    background: rgba(255, 255, 255, 0.3);"
        "}"
        "QComboBox QAbstractItemView {"
        "    background: rgba(0, 0, 0, 0.2);"
        "    border: none;"
        "    color: white;"
        "    border-radius: 21px;"
        "    padding: 5px;"
        "    selection-background-color: rgba(255, 255, 255, 0.2);"
        "}"
    );
    
    // Brightness label
    brightnessLabel = new QLabel("Brightness", displayBox);
    QFont brightFont("ElysiaOSNew", 14);
    brightnessLabel->setFont(brightFont);
    brightnessLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
    brightnessLabel->move(220, 300);
    
    // Brightness slider
    brightnessSlider = new QSlider(Qt::Horizontal, displayBox);
    brightnessSlider->setGeometry(100, 340, 400, 30);
    brightnessSlider->setMinimum(1);
    brightnessSlider->setMaximum(100);
    brightnessSlider->setStyleSheet(
        "QSlider::horizontal {"
        "    background: transparent;"
        "    border: none;"
        "    padding: 0px;"
        "    margin: 0px;"
        "}"
        "QSlider::groove:horizontal {"
        "    height: 8px;"
        "    background: transparent;"
        "    border: none;"
        "    padding: 0px;"
        "    margin: 0px;"
        "}"
        "QSlider::handle:horizontal {"
        "    background: pink;"
        "    border: 1px solid white;"
        "    width: 18px;"
        "    margin: -6px 0;"
        "    border-radius: 9px;"
        "}"
        "QSlider::sub-page:horizontal {"
        "    background: rgba(255, 255, 255, 0.25);"
        "    border-radius: 5px;"
        "}"
        "QSlider::add-page:horizontal {"
        "    background: rgba(255, 255, 255, 0.05);"
        "    border-radius: 5px;"
        "}"
    );
    connect(brightnessSlider, &QSlider::valueChanged, this, &DisplayManager::setBrightness);
    
    // Apply button
    applyBtn = new QPushButton("Apply Changes", displayBox);
    applyBtn->setGeometry(200, 400, 200, 50);
    applyBtn->setStyleSheet(
        "QPushButton {"
        "    background-color: #e5a7c6;"
        "    color: white;"
        "    font-weight: bold;"
        "    border: none;"
        "    padding: 6px 12px;"
        "    letter-spacing: 1px;"
        "    border-radius: 10px;"
        "    font-size: 16px;"
        "}"
        "QPushButton:hover {"
        "    background-color: #edcee3;"
        "}"
    );
    connect(applyBtn, &QPushButton::clicked, this, &DisplayManager::applyDisplayChange);
}

QString DisplayManager::detectThemeImage() {
    QString basePath = QDir::homePath() + "/.config/Elysia/assets/";
    try {
        QProcess process;
        process.start("gsettings", QStringList() << "get" << "org.gnome.desktop.interface" << "color-scheme");
        if (process.waitForFinished()) {
            QString output = QString::fromUtf8(process.readAllStandardOutput()).trimmed();
            if (output.contains("prefer-light")) {
                return basePath + "assets/light1.png";
            } else {
                return basePath + "assets/dark1.png";
            }
        }
    } catch (...) {
        // Fall through to default
    }
    return basePath + "assets/light1.png";
}

QStringList DisplayManager::getMonitorModes() {
    QStringList modes;
    try {
        QProcess process;
        process.start("hyprctl", QStringList() << "monitors");
        if (process.waitForFinished()) {
            QString output = QString::fromUtf8(process.readAllStandardOutput());
            QRegularExpression re(R"(\d{3,4}x\d{3,4}@\d{2,3})");
            QRegularExpressionMatchIterator it = re.globalMatch(output);
            while (it.hasNext()) {
                QString mode = it.next().captured(0);
                if (!modes.contains(mode)) {
                    modes.append(mode);
                }
            }
        }
    } catch (...) {
        // Return empty list on error
    }
    std::sort(modes.begin(), modes.end());
    return modes;
}

void DisplayManager::updateDisplayDropdown() {
    QStringList modes = getMonitorModes();
    QStringList currentItems;
    for (int i = 0; i < displayDropdown->count(); ++i) {
        currentItems.append(displayDropdown->itemText(i));
    }
    
    if (modes != currentItems) {
        displayDropdown->clear();
        displayDropdown->addItems(modes);
    }
    
    QString currentMode;
    QString configPath = QDir::homePath() + "/.config/hypr/hyprland.conf";
    QFile file(configPath);
    if (file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        QTextStream in(&file);
        while (!in.atEnd()) {
            QString line = in.readLine();
            if (line.startsWith("monitor") && line.contains("@")) {
                QRegularExpression re(R"(\d{3,4}x\d{3,4}@\d{2,3})");
                QRegularExpressionMatch match = re.match(line);
                if (match.hasMatch()) {
                    currentMode = match.captured(0);
                }
                QStringList parts = line.split(QRegularExpression("\\s+"));
                if (parts.size() > 1) {
                    monitorName = parts[1];
                }
                break;
            }
        }
        file.close();
    }
    
    if (!currentMode.isEmpty() && modes.contains(currentMode)) {
        displayDropdown->setCurrentText(currentMode);
    }
}

void DisplayManager::updateThemeImage() {
    QString imagePath = detectThemeImage();
    QPixmap pixmap(imagePath);
    desktopImageLabel->setPixmap(pixmap.scaled(300, 500, Qt::KeepAspectRatio, Qt::SmoothTransformation));
}

void DisplayManager::setBrightness(int value) {
    try {
        QProcess::startDetached("brightnessctl", QStringList() << "set" << QString::number(value) + "%");
    } catch (...) {
        QMessageBox::warning(qobject_cast<QWidget*>(parent()), "Error", "Failed to set brightness");
    }
}

void DisplayManager::updateBrightnessSlider() {
    try {
        QProcess process;
        process.start("brightnessctl", QStringList() << "g");
        if (process.waitForFinished()) {
            QString output = QString::fromUtf8(process.readAllStandardOutput()).trimmed();
            
            QProcess maxProcess;
            maxProcess.start("brightnessctl", QStringList() << "m");
            if (maxProcess.waitForFinished()) {
                QString maxOutput = QString::fromUtf8(maxProcess.readAllStandardOutput()).trimmed();
                
                bool ok1, ok2;
                int current = output.toInt(&ok1);
                int max = maxOutput.toInt(&ok2);
                
                if (ok1 && ok2 && max > 0) {
                    int percent = (current * 100) / max;
                    if (brightnessSlider->value() != percent) {
                        brightnessSlider->setValue(percent);
                    }
                }
            }
        }
    } catch (...) {
        brightnessSlider->setValue(50);
    }
}

QPair<int, QStringList> DisplayManager::getCurrentMonitorLine() {
    QString path = QDir::homePath() + "/.config/hypr/hyprland.conf";
    QFile file(path);
    QStringList lines;
    
    if (file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        QTextStream in(&file);
        while (!in.atEnd()) {
            lines.append(in.readLine());
        }
        file.close();
    }
    
    for (int i = 0; i < lines.size(); ++i) {
        if (lines[i].startsWith("monitor") && lines[i].contains("@")) {
            return qMakePair(i, lines);
        }
    }
    
    return qMakePair(-1, lines);
}

void DisplayManager::applyDisplayChange() {
    QString newMode = displayDropdown->currentText();
    QPair<int, QStringList> result = getCurrentMonitorLine();
    int index = result.first;
    QStringList lines = result.second;
    
    if (index == -1) {
        QMessageBox::warning(qobject_cast<QWidget*>(parent()), "Error", "Could not find monitor config line.");
        return;
    }
    
    QRegularExpression re(R"(\d{3,4}x\d{3,4}@\d{2,3})");
    lines[index] = lines[index].replace(re, newMode);
    
    QString configPath = QDir::homePath() + "/.config/hypr/hyprland.conf";
    QFile file(configPath);
    if (file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        QTextStream out(&file);
        for (const QString &line : lines) {
            out << line << "\n";
        }
        file.close();
    } else {
        QMessageBox::warning(qobject_cast<QWidget*>(parent()), "Error", "Failed to write file");
        return;
    }
    
    QMessageBox::information(qobject_cast<QWidget*>(parent()), "Success", 
                           QString("Display mode changed to %1.\nRestart Hyprland for changes to apply.").arg(newMode));
}

void DisplayManager::show() {
    updateDisplayDropdown();
    updateBrightnessSlider();
    updateThemeImage();
    
    desktopImageLabel->setVisible(true);
    desktopLabel->setVisible(true);
    displayBox->setVisible(true);
    
    refreshTimer->start();
}

void DisplayManager::hide() {
    refreshTimer->stop();
    displayBox->setVisible(false);
    desktopImageLabel->setVisible(false);
    desktopLabel->setVisible(false);
}

void DisplayManager::autoRefresh() {
    if (displayBox->isVisible()) {
        updateBrightnessSlider();
        updateDisplayDropdown();
        updateThemeImage();
    }
} 